npm/pkg/dataplane/ipsets/ipsetmanager.go

```diff
@@ -23,261 +25,346 @@ func (iMgr *IPSetManager) exists(name string) bool {
+			npmErrorString = npmerrors.AddNetPolReference
+		}
+		return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s does not exist", setName))
+	}
+
+	set := iMgr.setMap[setName]
+	wasInKernel := set.shouldBeInKernel()
+	set.addReference(referenceName, referenceType)
+	if !wasInKernel {
+		iMgr.modifyCacheForKernelCreation(set.Name)
+
+		// if set.Kind == HashSet, then this for loop will do nothing
+		for _, member := range set.MemberIPSets {
+			iMgr.incKernelReferCountAndModifyCache(member)
+		}
+	}
+	return nil
+}
+
+func (iMgr *IPSetManager) incKernelReferCountAndModifyCache(member *IPSet) {
+	wasInKernel := member.shouldBeInKernel()
+	member.incKernelReferCount()
+	if !wasInKernel {
+		iMgr.modifyCacheForKernelCreation(member.Name)
 	}
+}
 
-	// If set is not referenced in netpol then ignore the update
-	if len(set.NetPolReference) == 0 && len(set.SelectorReference) == 0 {
-		return
+func (iMgr *IPSetManager) DeleteReference(setName, referenceName string, referenceType ReferenceType) error {
+	if !iMgr.exists(setName) {
+		npmErrorString := npmerrors.DeleteSelectorReference
+		if referenceType == NetPolType {
+			npmErrorString = npmerrors.DeleteNetPolReference
+		}
+		return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s does not exist", setName))
 	}
 
-	iMgr.dirtyCaches[set.Name] = struct{}{}
-	if set.Kind == ListSet {
-		// TODO check if we will need to add all the member ipsets
-		// also to the dirty cache list
+	set := iMgr.setMap[setName]
+	wasInKernel := set.shouldBeInKernel() // required because the set may have 0 references i.e. this reference doesn't exist
+	set.deleteReference(referenceName, referenceType)
+	if wasInKernel && !set.shouldBeInKernel() {
+		iMgr.modifyCacheForKernelRemoval(set.Name)
+
+		// if set.Kind == HashSet, then this for loop will do nothing
 		for _, member := range set.MemberIPSets {
-			iMgr.dirtyCaches[member.Name] = struct{}{}
+			iMgr.decKernelReferCountAndModifyCache(member)
 		}
 	}
+	return nil
+}
+
+func (iMgr *IPSetManager) decKernelReferCountAndModifyCache(member *IPSet) {
+	member.decKernelReferCount()
+	if !member.shouldBeInKernel() {
+		iMgr.modifyCacheForKernelRemoval(member.Name)
+	}
 }
 
 func (iMgr *IPSetManager) clearDirtyCache() {
-	iMgr.dirtyCaches = make(map[string]struct{})
+	iMgr.additionOrUpdateDirtyCache = make(map[string]struct{})
+	iMgr.deletionDirtyCache = make(map[string]struct{})
 }
 
-func (iMgr *IPSetManager) CreateIPSet(set *IPSet) error {
+func (iMgr *IPSetManager) CreateIPSet(setName string, setType SetType) error {
 	iMgr.Lock()
 	defer iMgr.Unlock()
-	return iMgr.createIPSet(set)
+	return iMgr.createIPSet(setName, setType)
 }
 
-func (iMgr *IPSetManager) createIPSet(set *IPSet) error {
-	// Check if the Set already exists
-	if iMgr.exists(set.Name) {
-		// ipset already exists
-		// we should calculate a diff if the members are different
-		return nil
+func (iMgr *IPSetManager) createIPSet(setName string, setType SetType) error {
```

Comment:
We had this smaller createIPSet so we can call it from other functions, if we are going to return an error when an ipset is missing, lets remove this function and consolidate into CreateIPSet